app/Services/TransitionProcessor.php

```diff
@@ -0,0 +1,293 @@
+<?php
+
+namespace App\Services;
+
+use App\Events\VoucherPaymentRequested;
+use App\Http\Controllers\API\TraderController;
+use App\StateToken;
+use App\Trader;
+use App\Voucher;
+use Carbon\Carbon;
+use Exception;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Log;
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Lock\Store\SemaphoreStore;
+
+class TransitionProcessor
+{
+    public array $responses = [
+        'success_add' => [],
+        'success_reject' => [],
+        'own_duplicate' => [],
+        'other_duplicate' => [],
+        'invalid' => [],
+        'failed_reject' => [],
+        'undelivered' => [],
+    ];
+
+    public array $vouchers_for_payment = [];
+
+    private Trader $trader;
+
+    private Collection $vouchers;
+
+    private string $transition;
+
+    public function __construct(trader $trader, string $transition)
+    {
+        $this->transition = $transition;
+        $this->trader = $trader;
+    }
+
+    public function handle(array $voucherCodes): array
+    {
+        // set a lock, to prevent double submits
+        $store = new SemaphoreStore();
+        $factory = new LockFactory($store);
+        $lock = $factory->createLock('transition');
+
+        if ($lock->acquire()) {
+            // get and the available vouchers
+            $this->vouchers = Voucher::findByCodes($voucherCodes);
+
+            // Get the ones not in that list - they are bad codes.
+            // We need to re-key the array here because otherwise the json response will return object for non 0 starting.
+            $this->responses['invalid'] = array_values(
+                array_diff(
+                    $voucherCodes,
+                    $this->vouchers->pluck('code')->toArray()
+                )
+            );
+
+            switch ($this->transition) {
+                case 'collect' :
+                    $this->handleCollect();
+                    break;
+                case 'confirm':
+                    $this->handleConfirm();
+                    break;
+                case 'reject':
+                    $this->handleReject();
+                    break;
+                default:
+                    $this->handleDefault();
+            }
+
+            $lock->release();
+        } else {
+            Log::info('No lock for me!');
+            $this->responses['own_duplicate'][] = '000000';
+        }
+        return $this->responses;
+    }
+
+    /**
+     * @return void
+     */
+    public function handleCollect(): void
+    {
+        // Fetch the date we start to care about deliveries
+        $collect_delivery_date = Carbon::parse(config('arc.first_delivery_date'));
+        $transition = $this->transition;
+
+        sleep(10);
```

Comment:
Is this still needed now there's a lock?

Reply:
doh!